velox/common/hyperloglog/DenseHll.cpp

```diff
@@ -558,46 +558,224 @@ void DenseHll::mergeWith(const char* serialized) {
+  // One register will store values with odd indices (0, 2, 4...). The other
+  // register will store values with even indices (1, 3, 5...).
+
+  // Load deltas with even indices into SIMD register.
+  auto loadEven = [&](const int8_t* deltas) {
+    auto batch = xsimd::load_unaligned(deltas);
+    batch = xsimd::kernel::bitwise_rshift(batch, 4, xsimd::default_arch{});
+    return xsimd::bitwise_and(batch, bucketMaskBatch);
+  };
+
+  // Load deltas with odd indices into SIMD register.
+  auto loadOdd = [&](const int8_t* deltas) {
+    auto batch = xsimd::load_unaligned(deltas);
+    return xsimd::bitwise_and(batch, bucketMaskBatch);
+  };
+
+  // Count number of zeros in a SIMD register.
+  auto countZeros = [&](const xsimd::batch<int8_t>& batch) {
+    auto zerosBitmask = xsimd::eq(batch, zeroBatch).mask();
+    return bits::countBits(&zerosBitmask, 0, batchSize);
+  };
+
+  // Given two SIMD registers of deltas, converts deltas to values by adding
+  // baselines and returns their max along with a bitmask that has bits set for
+  // entries that may have an overflow.
+  auto processBatch = [&](xsimd::batch<int8_t>& batch,
+                          xsimd::batch<int8_t>& otherBatch) {
+    auto overflows = xsimd::eq(batch, maxDeltaBatch).mask();
+    batch += baselineBatch;
+
+    overflows |= xsimd::eq(otherBatch, maxDeltaBatch).mask();
+    otherBatch += otherBaselineBatch;
+
+    // Compute max.
+    auto maxBatch = xsimd::max(batch, otherBatch);
+    maxBatch -= newBaselineBatch;
+
+    return std::pair{maxBatch, overflows};
+  };
+
+  auto processOverflow =
+      [&](int8_t delta1, int8_t delta2, int bucket, int shift) {
+        auto [newValue, overflowEntry] =
+            computeNewValue(delta1, delta2, bucket, other);
+
+        int8_t newDelta = newValue - newBaseline;
+
+        return updateOverflow(bucket, overflowEntry, newDelta);
+      };
+
+  std::vector<int8_t> evenTemp(batchSize * 2);
+  std::vector<int8_t> oddTemp(batchSize * 2);
+
+  int32_t baselineCount = 0;
+  for (int i = 0; i < deltas_.size(); i += batchSize) {
+    // Process values in even indices first.
+    auto evenBatch = loadEven(deltas_.data() + i);
+    auto otherEvenBatch = loadEven(other.deltas + i);
+
+    auto [evenMaxBatch, evenOverflows] =
+        processBatch(evenBatch, otherEvenBatch);
+
+    baselineCount += countZeros(evenMaxBatch);
+
+    // Process values in odd indices.
+    auto oddBatch = loadOdd(deltas_.data() + i);
+    auto otherOddBatch = loadOdd(other.deltas + i);
+
+    auto [oddMaxBatch, oddOverflows] = processBatch(oddBatch, otherOddBatch);
+
+    baselineCount += countZeros(oddMaxBatch);
+
+    // Combine even and odd batches. Shift even batch left by 4 bits, then OR
+    // with odd batch.
+    auto combinedBatch =
+        xsimd::kernel::bitwise_lshift(evenMaxBatch, 4, xsimd::default_arch{});
+    combinedBatch = xsimd::bitwise_or(combinedBatch, oddMaxBatch);
+
+    xsimd::store_unaligned(deltas_.data() + i, combinedBatch);
```

Comment:
A different approach would be creating temporary memory for `deltas_`:

```c++
if (evenOverflows == 0 && oddOverflows == 0) {
  xsimd::store_unaligned(deltas_.data() + i, combinedBatch);
  continue;
}
int8_t deltas[batchSize];
xsimd::store_unaligned(deltas, combinedBatch);
/// ...
deltas[index] = (newDelta << 4) | (slot1 & kBucketMask);
/// ...
memcpy(deltas_.data() + i, deltas, batchSize);
```
This way we reduce the copy from SIMD registers to memory from 4 to 1 load + 1 store.